api/buildings/models.py

```diff
@@ -248,3 +257,72 @@ class Meta:
 
     def __str__(self):
         return self.name
+
+
+class ApprovedImage(models.Manager):
+    def get_queryset(self):
+        return super().get_queryset().filter(status=ImageFile.ACCEPTED)
+
+
+class ImageFile(models.Model):
+    PENDING = 0
+    ACCEPTED = 1
+    REJECTED = -1
+
+    IMAGE_STATUS_CHOICES = [
+        (PENDING, _("Pending")),
+        (ACCEPTED, _("Accepted")),
+        (REJECTED, _("Rejected")),
+    ]
+
+    def set_images_limit(building):
+        if ImageFile.objects.filter(building_id=building).count() >= settings.ALLOWED_IMAGES_LIMIT:
+            raise ValidationError("Image limit for building is reached (%s)" % settings.ALLOWED_IMAGES_LIMIT)
+
+    def image_thumb(self):
+        return mark_safe(
+            '<a href={0}><img src="{0}" url width="50" height="50" /></a>'.format(
+                os.path.join(settings.MEDIA_URL, str(self.image))
+            )
+        )
+
+    image_thumb.short_description = "Thumbnail"
+
+    building = models.ForeignKey(Building, validators=(set_images_limit,), on_delete=models.CASCADE)
+    name = models.CharField(_("name"), max_length=255)
+    image_type = models.CharField(_("type"), max_length=255, choices=settings.ACCEPTED_IMAGE_TYPES)
+    image = models.ImageField(default="Add image file", upload_to="images/")
+    status = models.SmallIntegerField(_("status"), default=PENDING, choices=IMAGE_STATUS_CHOICES, db_index=True)
+
+    def __str__(self):
+        return self.name
+
+    def save(self):
+        # Opening the uploaded image
+        im = PIL.Image.open(self.image)
+
+        output = BytesIO()
+
+        # TODO: do we want to resize it in any way ?
+        # Resize/modify the image
+        # im = im.resize((500, 500))
+
+        # after modifications, save it to the output
+        im.save(
```

Comment:
The image processing should take place only if the image file was changed, not on each model save.

Reply:
Good catch, I've updated the code to make this possible.